public/app/features/explore/state/epics/runRequestEpic.ts

```diff
@@ -0,0 +1,205 @@
+  runQueriesBatchAction,
+  RunQueriesBatchPayload,
+  queryStartAction,
+  historyUpdatedAction,
+  processQueryResultsAction,
+  stateSaveAction,
+  processQueryErrorsAction,
+  changeRangeAction,
+  limitMessageRatePayloadAction,
+  changeLoadingStateAction,
+} from '../actionTypes';
+import { buildQueryTransaction, updateHistory } from '../../../../core/utils/explore';
+import { PanelQueryState } from '../../../dashboard/state/PanelQueryState';
+import { DataQueryResponseData } from '@grafana/ui/src/types/datasource';
+import { ExploreId, ExploreItemState } from 'app/types';
+import { LoadingState, DataFrame, AbsoluteTimeRange, dateMath } from '@grafana/data';
+import { PanelData } from '@grafana/ui/src/types/panel';
+import { isString } from 'lodash';
+
+const publishActions = (outerObservable: Subscriber<unknown>, actions: Array<ActionOf<any>>) => {
+  for (const action of actions) {
+    outerObservable.next(action);
+  }
+};
+
+interface ProcessResponseConfig {
+  exploreId: ExploreId;
+  exploreItemState: ExploreItemState;
+  datasourceId: string;
+  now: number;
+  loadingState: LoadingState;
+  series?: DataQueryResponseData[];
+  delta?: DataFrame[];
+}
+
+const processResponse = (config: ProcessResponseConfig) => {
+  const { exploreId, exploreItemState, datasourceId, now, loadingState, series, delta } = config;
+  const { queries, history } = exploreItemState;
+  const latency = Date.now() - now;
+
+  // Side-effect: Saving history in localstorage
+  const nextHistory = updateHistory(history, datasourceId, queries);
+  return [
+    historyUpdatedAction({ exploreId, history: nextHistory }),
+    processQueryResultsAction({ exploreId, latency, datasourceId, loadingState, series, delta }),
+    stateSaveAction(),
+  ];
+};
+
+interface ProcessErrorConfig {
+  exploreId: ExploreId;
+  datasourceId: string;
+  error: any;
+}
+
+const processError = (config: ProcessErrorConfig) => {
+  const { exploreId, datasourceId, error } = config;
+
+  return [processQueryErrorsAction({ exploreId, response: error, datasourceId })];
+};
+
+export const runRequestEpic: Epic<ActionOf<any>, ActionOf<any>, StoreState> = (
+  action$,
+  state$,
+  { getQueryResponse }
+) => {
+  return action$.ofType(runQueriesBatchAction.type).pipe(
+    mergeMap((action: ActionOf<RunQueriesBatchPayload>) => {
+      const { exploreId, queryOptions } = action.payload;
+      const exploreItemState = state$.value.explore[exploreId];
+      const { datasourceInstance, queries, queryIntervals, range, scanning } = exploreItemState;
+
+      // Create an observable per run queries action
+      // Within the observable create two subscriptions
+      // First subscription: 'querySubscription' subscribes to the call to query method on datasourceinstance
+      // Second subscription: 'streamSubscription' subscribes to events from the query methods observer callback
+      const observable: Observable<ActionOf<any>> = new Observable(outerObservable => {
```

Comment:
When I compare the end result of `runRequestEpic` with `runQueriesBatchEpic` I find them very similar and I can't see the advantage of `PanelQueryState`. I probably don't understand how to use it in the context of an epic yet, so I will do another take on this next week.

Seems like `runQueriesBatchEpic` === `PanelQueryState` except for the explore specific cancellations in takeUntil
https://github.com/grafana/grafana/blob/e50a75f25f0f9e0374af110c14198cb4ce643a97/public/app/features/explore/state/epics/runQueriesBatchEpic.ts#L197

Reply:
> Seems like runQueriesBatchEpic === PanelQueryState except for the explore specific cancellations in takeUntil

I think it is more like `runQueriesBatchEpic ~= PanelQueryState` -- and in an effort to make sure we know it is `===` it would be best to reuse the code.

The differences happen with 
* multiple streams
* queries that return multiple responses with different keys
* mixed queries... then things get really weird.

Check the test in [panelQueryState.test.ts](https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryState.test.ts#L102) after line 100 for a few cases